paasta_tools/autoscaling_lib.py

```diff
@@ -569,104 +586,166 @@ def sort_slaves_to_kill(mesos_state, pool='default'):
-                for slave_pid, slave_data in instances_to_kill.items():
-                    instance_id = slave_data['instance_id']
-                    if not instance_id:
-                        log.warning("Didn't find instance ID for slave: {0}. Skipping terminating".format(slave_pid))
-                        continue
-                    is_ready_to_kill = True
-                    # is_ready_to_kill = paasta_maintenance.something(slave['pid'])
-                    if is_ready_to_kill:
-                        log.info("TERMINATING: {0}".format(instance_id))
-                        try:
-                            ec2_client.terminate_instances(InstanceIds=[instance_id], DryRun=dry_run)
-                        except ClientError as e:
-                            if e.response['Error'].get('Code') == 'DryRunOperation':
-                                pass
-                            else:
-                                raise
-                        instances_to_kill.pop(slave_pid)
-                    else:
-                        log.debug("Instance {0}: NOT ready to kill".format(instance_id))
+            while True:
+                instance_id = slave['instance_id']
+                if not instance_id:
+                    log.warning("Didn't find instance ID for slave: {0}. Skipping terminating".format(slave['pid']))
+                    continue
+                is_ready_to_kill = True
+                # is_ready_to_kill = paasta_maintenance.something(slave['pid'])
+                if is_ready_to_kill:
+                    log.info("TERMINATING: {0}".format(instance_id))
+                    try:
+                        ec2_client.terminate_instances(InstanceIds=[instance_id], DryRun=dry_run)
+                    except ClientError as e:
+                        if e.response['Error'].get('Code') == 'DryRunOperation':
+                            pass
+                        else:
+                            raise
+                    break
+                else:
+                    log.debug("Instance {0}: NOT ready to kill".format(instance_id))
                 log.debug("Waiting 5 seconds and then checking again")
                 time.sleep(5)
     except TimeoutError:
         log.error("Timed out after {0} waiting to drain {1}, now terminating anyway".format(CLUSTER_DRAIN_TIMEOUT,
-                                                                                            instances_to_kill))
+                                                                                            slave['pid']))
         try:
-            ec2_client.terminate_instances(InstanceIds=instances_to_kill, DryRun=dry_run)
+            ec2_client.terminate_instances(InstanceIds=instance_id, DryRun=dry_run)
         except ClientError as e:
             if e.response['Error'].get('Code') == 'DryRunOperation':
                 pass
             else:
                 raise
 
 
+def get_instance_type_weights(sfr_id):
```

Comment:
Yeah, I think this is a good shout. I may still need to call multiple times to get the state when I check in `set_spot_fleet_request_capacity` but I could minimize it elsewhere. I'll take a look 👍